Intersect.Server.Core/Entities/Player.cs

```diff
@@ -1275,51 +1276,100 @@ public void SetLevel(int level, bool resetExperience = false)
         PacketSender.SendExperience(this);
     }
 
-    public void LevelUp(bool resetExperience = true, int levels = 1)
+    public void AddLevels(int levels = 1, bool resetExperience = true)
     {
-        var messages = new List<string>();
-        if (Level < Options.Instance.Player.MaxLevel)
+        if (levels > 0)
         {
-            for (var i = 0; i < levels; i++)
+            while (levels > 0)
             {
-                SetLevel(Level + 1, resetExperience);
-
-                //Let's pull up class - leveling info
-                var classDescriptor = ClassBase.Get(ClassId);
-                if (classDescriptor?.Spells == null)
+                if (Level >= Options.Instance.Player.MaxLevel)
                 {
-                    continue;
+                    break;
                 }
 
-                foreach (var spell in classDescriptor.Spells)
+                SetLevel(Level + 1, resetExperience);
+                StartCommonEventsWithTrigger(CommonEventTrigger.LevelUp);
+
+                PacketSender.SendChatMsg(
+                    this,
+                    Strings.Player.LevelUp.ToString(Level),
+                    ChatMessageType.Experience,
+                    CustomColors.Combat.LevelUp,
+                    Name
+                );
+
+                if (ClassBase.TryGet(ClassId, out var classDescriptor) && classDescriptor?.Spells != default)
                 {
-                    if (spell.Level != Level)
+                    foreach (var spell in classDescriptor.Spells)
                     {
-                        continue;
+                        if (spell.Level != Level)
+                        {
+                            continue;
+                        }
+
+                        var spellInstance = new Spell(spell.Id);
+                        _ = TryTeachSpell(spellInstance, true);
```

Comment:
Where's the message for learning spells? Why is there only a message for forgetting them in this method?

Reply:
It is within the method itself (TryTeachSpell), when I tested it, before recording the video, the messages when learning magic when passing a level were duplicated, as this method already sends a message when teaching

Reply:
fixed now